rllib/algorithms/algorithm.py

```diff
@@ -1595,35 +1599,98 @@ def add_policy(
         Returns:
             The newly added policy (the copy that got added to the local
             worker).
+
+        Raises:
+            ValueError: If both `policy_cls` AND `policy` are provided.
+            KeyError: If the given `policy_id` already exists in this Algorithm.
         """
+        local_worker = self.workers.local_worker()
 
-        kwargs = dict(
-            policy_id=policy_id,
-            policy_cls=policy_cls,
-            observation_space=observation_space,
-            action_space=action_space,
-            config=config,
-            policy_state=policy_state,
-            policy_mapping_fn=policy_mapping_fn,
-            policies_to_train=list(policies_to_train) if policies_to_train else None,
-        )
+        if policy_id in local_worker.policy_map:
+            raise KeyError(
+                f"Policy ID '{policy_id}' already exists in policy map! "
+                "Make sure you use a Policy ID that has not been taken yet."
+                " Policy IDs that are already in your policy map: "
+                f"{list(local_worker.policy_map.keys())}"
+            )
+
+        if policy_cls is not None and policy is not None:
```

Comment:
probably need to check if both are None?

Reply:
great catch! :)

Reply:
done